linalg/slepc.cpp

```diff
@@ -0,0 +1,342 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "../config/config.hpp"
+
+#ifdef MFEM_USE_MPI
+#ifdef MFEM_USE_PETSC
+#ifdef MFEM_USE_SLEPC
+
+#include "linalg.hpp"
+
+#include "slepc.h"
+
+// Error handling
+// Prints SLEPc's stacktrace and then calls MFEM_ABORT
+// We cannot use PETSc's CHKERRQ since it returns a PetscErrorCode
+#define PCHKERRQ(obj,err) do {                                                   \
+     if ((err))                                                                  \
+     {                                                                           \
+        PetscError(PetscObjectComm((PetscObject)(obj)),__LINE__,_MFEM_FUNC_NAME, \
+                   __FILE__,(err),PETSC_ERROR_REPEAT,NULL);                      \
+        MFEM_ABORT("Error in SLEPc. See stacktrace above.");                     \
+     }                                                                           \
+  } while(0);
+#define CCHKERRQ(comm,err) do {                                \
+     if ((err))                                                \
+     {                                                         \
+        PetscError(comm,__LINE__,_MFEM_FUNC_NAME,              \
+                   __FILE__,(err),PETSC_ERROR_REPEAT,NULL);    \
+        MFEM_ABORT("Error in SLEPc. See stacktrace above.");   \
+     }                                                         \
+  } while(0);
+
+
+static PetscErrorCode ierr;
+
+namespace mfem
+{
+void MFEMInitializeSlepc()
+{
+   MFEMInitializeSlepc(NULL,NULL,NULL,NULL);
+}
+
+void MFEMInitializeSlepc(int *argc,char*** argv)
+{
+   MFEMInitializeSlepc(argc,argv,NULL,NULL);
+}
+
+void MFEMInitializeSlepc(int *argc,char ***argv,const char rc_file[],
+                         const char help[])
+{
+   ierr = SlepcInitialize(argc,argv,rc_file,help);
+   MFEM_VERIFY(!ierr,"Unable to initialize SLEPc");
+}
+
+void MFEMFinalizeSlepc()
+{
+   ierr = SlepcFinalize();
+   MFEM_VERIFY(!ierr,"Unable to finalize SLEPc");
+}
+
+
+SlepcEigenSolver::SlepcEigenSolver(MPI_Comm comm, const std::string &prefix,
+                                   bool wrap)
+{
+   clcustom = false;
+   _tol = PETSC_DEFAULT;
```

Comment:
I do not see the reason why you want to store `_tol`, `_max_its` and `_num_conv` in the class since you are not using them.

Reply:
It seems like `EPSSetTolerances` is different from e.g. `KSPSetTolerances`, in that if you pass PETSC_DEFAULT it will reset that parameter to default, instead of ignoring it. Because of this we need to keep an internal state if we want to change only one at a time. Do you prefer that I simply wrap `EPSSetTolerances`?

Reply:
In principle, the MFEM API should only be a proxy for EPS API.
You can call EPSGetTolerances inside MFEM API to preserve any already set value. 
If we want to keep SetTol and SetMaxIt separate, what should we do? @joseeroman?

Reply:
If you want to change only one of the arguments, call `EPSGetTolerances` first and then `EPSSetTolerances` with one value modified.

Reply:
And this will work even if the parameter has not been set before? Keeping the state of the EPS solver intact and not having any unexpected side-effect? Should we call EPSSetTolerance(eps,PETSC_DEFAULT,PETSC_DEFAULT) at creation time to avoid this?

Reply:
Now that I look at the code, the two parameters have different behaviour, the tolerance is set initially to `PETSC_DEFAULT` but the maximum iterations have the value 0 to indicate uninitialized. So `EPSGetTolerances` followed by `EPSSetTolerances` will work for the former but not the latter. Do you want me to fix it?

Reply:
You mean you need to fix it in SLEPc? If zero is a valid value for maxit, then what's wrong about calling `EPSSetTolerance(eps,PETSC_DEFAULT,PETSC_DEFAULT)` at object creation and then use the get-first-then-set approach for the tolerances?

Reply:
Yes, I mean fixing in SLEPc.
No, maxit=zero is not allowed in `EPSSetTolerances`, that is the problem.

Reply:
Oh, I see now. If 0 will always indicate uninitialized maxits in SLEPc, we can do

```
EPSGetTolerances(eps,&t,&m)
if (m==0) m = PETSC_DECIDE;
EPSSetTolerances(eps,t,m)
```
We don't need to by tight to a specific version of SLEPc having this fix.

Reply:
Correct. Anyway, I will fix it in slepc-maint

Reply:
Thanks for the clarifications, I will add that so it will work around the previous versions of SLEPc.